### ISSUE TITLE
ABI Checker

### ISSUE THREAD

**Issue by phracek** [phracek/rebase-helper#101]
Implementation ABI checker for binaries.

We should implement ABI checker so that user is inform what was changed from ABI point of view. Discussion is welcome.

**Comment by dodjiseketeli** [comment]
It'd be great to use the [abidiff tool] (https://sourceware.org/libabigail/manual/abidiff.html) to provide detailed reports about ABI changes in rebase-helper.

Today, abidiff works on two ELF binaries.  We'd ideally need that project to provide a wrapper that compares two RPM packages and provide a consolidated report of the ABI changes for the binaries included in.

**Comment by sinnykumari** [comment]
Yes, we can add a new tool in libabigail called pkgabidiff (tentative name)  which will run between two packages and provides ABI diff results. rebase-helper can easily invoke pkgabidiff binary along with required arguments and can get desired ABI checks feature.

**Comment by phracek** [comment]
Well, what is difference between libabigail and abidiff tool?
Do you plan to release a wrapper also to abidiff tool?

**Comment by phracek** [comment]
Bugzilla request in libabigail https://sourceware.org/bugzilla/show_bug.cgi?id=18426

**Comment by dodjiseketeli** [comment]
*Libabigail* is a project to write reusable parts to handle ABI related things in ELF binaries.  It's also the name of a low level software library that knows how to read ELF libraries along with their debug information to analyze and compare their ABIs. On top of the **Libabigail library**, the **abidiff** tool was written in the **Libabigail project** to compare the ABIs of two ELF libraries and emit a text report describing their differences.

**Comment by phracek** [comment]
https://sourceware.org/libabigail/manual/abipkgdiff.html

**Comment by dodjiseketeli** [comment]
So, the abipkgdiff tool is now available in a Rawhide and Fedora 22 (among others).  The Fedora 22 package is available at http://koji.fedoraproject.org/koji/buildinfo?buildID=670154 and should hit a mirror near you soon.

Cheers.

**Comment by phracek** [comment]
Would it be possible to add an option for computer readable format?
Current output is more or less for humans.

Otherwise abipkgdiff works fine.

**Comment by dodjiseketeli** [comment]
We can add such an output, yes.  Actually, I have started to work on an XML format for the result of the ABI comparison.  I have stopped working on that due to more pressing matters like ... well, abipkgdiff and other things :-)

The work has started in a branch at https://sourceware.org/git/gitweb.cgi?p=libabigail.git;a=shortlog;h=refs/heads/xml-report.

The format I am defining is quite big, because as you can imagine, because the different kinds of changes that libabigail recognises is quite big.  To have an idea of the schema of the format I have started to devise, you can look at https://sourceware.org/git/gitweb.cgi?p=libabigail.git;a=blob_plain;f=doc/manuals/ifac-schema.rng;hb=refs/heads/xml-report.

What kind of computer readable format do you have in mind?

And is there a particular use case you want to address?

**Comment by phracek** [comment]
Issue is fixed by commit https://github.com/phracek/rebase-helper/commit/b4aab02eba0ae8fb8a9e10a5d57bd4b4d7645bb1

**Comment by phracek** [comment]
I create a issues for better tracking.https://github.com/phracek/rebase-helper/issues/113

**Comment by dodjiseketeli** [comment]
Hey Petr,

Thanks for this commit!  I hope using abipkgdiff wasn't too cumbersome for you.

I'd like to give you a heads-up about a commit change in the way you handle the exit code of the abipkgdiff code program, though.

The meaning of the exit code is going to change in some package updates that are probably going to take place in some few days.  Basically, zero is still going to mean the two ABIs are identical.

But non-zero is going to be more than just 1.  This change has been committed to the code repository of libabigail, but the Fedora packages haven't been updated yet.

 As there are possibly different kinds of changes, and thus different kinds of non-zero value, I guess the best is to read the documentation that I have already updated at https://sourceware.org/libabigail/manual/abipkgdiff.html, where it says that the values of the exit code are the same as for the abidiff tool, and these values are described at https://sourceware.org/libabigail/manual/abidiff.html#abidiff-return-value-label.

In short, in commit https://github.com/phracek/rebase-helper/commit/b4aab02eba0ae8fb8a9e10a5d57bd4b4d7645bb1, rather than writting, in rebase-helper/checker.py:


+                if int(ret_code) != 0 and int(ret_code) != 1:
+                    raise RebaseHelperError(...)

You'd write something like:

+                ABIDIFF_ERROR = 1
+                ABIDIFF_USAGE_ERROR = 2

[...]

+                if ret_code & (ABIDIFF_ERROR | ABIDIFF_USAGE_ERROR) :
+                    raise RebaseHelperError(...)

Cheers,

**Comment by phracek** [comment]
Well, I have tried to do that but it seems that it returns me a problem.
Command like: 
abipkgdiff --d1 /home/phracek/work/Fedora/libpng15/rebase-helper-results/old/RPM/libpng15-debuginfo-1.5.22-1.fc21.x86_64.rpm --d2 /home/phracek/work/Fedora/libpng15/rebase-helper-results/new/RPM/libpng15-debuginfo-1.5.23-1.fc21.x86_64.rpm /home/phracek/work/Fedora/libpng15/rebase-helper-results/old/RPM/libpng15-1.5.22-1.fc21.x86_64.rpm /home/phracek/work/Fedora/libpng15/rebase-helper-results/new/RPM/libpng15-1.5.23-1.fc21.x86_64.rpm

returns a bunch of changes in API

but return value is 1 in bash of course:
~# echo $?
1
~#

**Comment by phracek** [comment]
https://github.com/phracek/rebase-helper/commit/0face45cdb628a0f89fe5a95139aa8b047cacf92

**Comment by sinnykumari** [comment]
Hi Petr,

It may be a bug in abipkgdiff. Can you please provide rpms on which you are getting wrong exit code? I looked into koji for exact builds but couldn't find here http://koji.fedoraproject.org/koji/packageinfo?packageID=16253 .

**Comment by dodjiseketeli** [comment]
Hello Petr,
The bitwise operations to detect if the first and second least significant bits (of value 1 and 2) are set, are relevant for libabigail from the master repository, at the moment.  We haven't updated the libabigail packages of Fedora to those versions of master yet.  I was just telling you in advance what was going to happen when we really do the update :)  So I think it's normal that you are seeing what you are seeing for now.

Cheers.

**Comment by phracek** [comment]
Hi,
All packages are here: 
https://phracek.fedorapeople.org/libpng15-1.5.22-1.fc21.x86_64.rpm
https://phracek.fedorapeople.org/libpng15-1.5.23-1.fc21.x86_64.rpm
https://phracek.fedorapeople.org/libpng15-debuginfo-1.5.22-1.fc21.x86_64.rpm
https://phracek.fedorapeople.org/libpng15-debuginfo-1.5.23-1.fc21.x86_64.rpm

Greetings

**Comment by sinnykumari** [comment]
Ah, I thought Peter is running abipkgdiff from master.
Anyway, I ran abipkgdiff (from master) on provided rpm and it return exit code 4 which signifies to ABI changes.
We will make a new Fedora build soon so that this feature is available through package installation as well.

**Comment by sinnykumari** [comment]
I have made another build (for rawhide, F23, F22, F21) of libabigail from latest master http://koji.fedoraproject.org/koji/packageinfo?packageID=20298 (NVR - libabigail-1.0-0.5.20150727gitf0d319a.*) . F21 and F22 build should be available should be pushed soon to stable. This build contains feature of  different exit status code for different type of abipkgdiff return.
